spark3/src/main/java/org/apache/iceberg/actions/Spark3CreateAction.java

```diff
@@ -0,0 +1,187 @@
+  private final Identifier destTableIdent;
+
+  // Optional Parameters for destination
+  private final Map<String, String> additionalProperties = Maps.newHashMap();
+
+  Spark3CreateAction(SparkSession spark, CatalogPlugin sourceCatalog, Identifier sourceTableIdent,
+                     CatalogPlugin destCatalog, Identifier destTableIdent) {
+
+    this.spark = spark;
+    this.sourceCatalog = checkSourceCatalog(sourceCatalog);
+    this.sourceTableIdent = sourceTableIdent;
+    this.destCatalog = checkDestinationCatalog(destCatalog);
+    this.destTableIdent = destTableIdent;
+
+    try {
+      this.sourceTable = (V1Table) this.sourceCatalog.loadTable(sourceTableIdent);
+      this.sourceCatalogTable = sourceTable.v1Table();
+    } catch (org.apache.spark.sql.catalyst.analysis.NoSuchTableException e) {
+      throw new NoSuchTableException("Cannot not find source table %s", sourceTableIdent);
+    } catch (ClassCastException e) {
+      throw new IllegalArgumentException(String.format("Cannot use non-v1 table %s as a source", sourceTableIdent), e);
+    }
+    validateSourceTable();
+
+    this.sourceTableLocation = CatalogUtils.URIToString(sourceCatalogTable.storage().locationUri().get());
+  }
+
+  @Override
+  public CreateAction withProperties(Map<String, String> properties) {
+    this.additionalProperties.putAll(properties);
+    return this;
+  }
+
+  @Override
+  public CreateAction withProperty(String key, String value) {
+    this.additionalProperties.put(key, value);
+    return this;
+  }
+
+  protected SparkSession spark() {
+    return spark;
+  }
+
+  protected String sourceTableLocation() {
+    return sourceTableLocation;
+  }
+
+  protected CatalogTable v1SourceTable() {
+    return sourceCatalogTable;
+  }
+
+  protected TableCatalog sourceCatalog() {
+    return sourceCatalog;
+  }
+
+  protected Identifier sourceTableIdent() {
+    return sourceTableIdent;
+  }
+
+  protected StagingTableCatalog destCatalog() {
+    return destCatalog;
+  }
+
+  protected Identifier destTableIdent() {
+    return destTableIdent;
+  }
+
+  protected Map<String, String> additionalProperties() {
+    return additionalProperties;
+  }
+
+  private void validateSourceTable() {
+    String sourceTableProvider = sourceCatalogTable.provider().get().toLowerCase(Locale.ROOT);
+    Preconditions.checkArgument(ALLOWED_SOURCES.contains(sourceTableProvider),
+        "Cannot create an Iceberg table from source provider: %s", sourceTableProvider);
+    Preconditions.checkArgument(!sourceCatalogTable.storage().locationUri().isEmpty(),
+        "Cannot create an Iceberg table from a source without an explicit location");
+  }
+
+  private StagingTableCatalog checkDestinationCatalog(CatalogPlugin catalog) {
+    Preconditions.checkArgument(catalog instanceof SparkSessionCatalog || catalog instanceof SparkCatalog,
+        "Cannot create Iceberg table in non Iceberg Catalog. Catalog %s was of class %s but %s or %s are required",
```

Comment:
nit: Do we add quotes around names or no? I am fine either way but having it consistent is probably good.

Reply:
we definitely add them in other places, I wasn't consistent in my reapplication though (I also think we aren't doing it consistently throughout the codebase but that's another story)

Reply:
I think that names should be internally quoted, not externally. So `["a", " ", "b"]` should be ```a.` `.b```.